README.md

```diff
@@ -13,64 +13,99 @@ pip install . # pip install ProtocolDataUnits
-| CRC Support                    | Yes                              | Yes                                  |
-| PDU Specifics                  |                                  |                                      |
-| Ethernet Frame                 | Yes                              | Yes (via EthernetFrame class)        |
-| Nested PDUs                    | Not Explicitly Mentioned         | Designed but not implemented         |
-| Byte Order Flexibility         | Yes (BIG_ENDIAN, LITTLE_ENDIAN)  | Yes                                  |
-| Variable Length Payload        | Yes                              | Yes                                  |
-| Additional Features            |                                  |                                      |
-| Pretty Printing                | Yes                              | Via __repr__ method                  |
-| Metadata Support               | Through PDUInfo                  | Yes, via metadata attribute          |
+`ProtocolDataUnits` is a Python toolset for encoding and decoding Protocol Data Units (PDUs). It includes the following features:
+
+- **Base PDU Definition**: Define the structure and format of your PDUs.
+- **PDU Encoding/Decoding**: Encode and decode PDUs with various field types.
+- **Nested PDU Support**: Supports PDUs within PDUs, allowing for complex data structures.
+- **CRC32 Checksum**: Automatically compute and validate CRC32 checksums for data integrity.
+- **Field Encoding/Decoding**: Supports a variety of field types including integers, floats, strings, and arrays.
+- **Byte Order Conversion**: Flexibly handle big-endian and little-endian byte orders.
+- **Metadata Storage**: Store additional metadata within PDUs.
+- **Stream Writing/Reading**: Efficiently write and read PDUs to and from streams.
+- **Variable Length Encoding/Decoding**: Handle fields with variable length data.
+- **Pretty Printing of PDUs**: Generate human-readable representations of PDUs.
+- **PDU Equality based on Fields**: Compare PDUs based on their field values.
+- **Serialization and Deserialization**: Serialize PDU definitions to JSON and deserialize them back to PDU objects.
 
 ## Usage
 
+### Creating PDU Formats
+
+You can create PDU formats using the user-friendly API:
+
 ```python
-from ProtocolDataUnits.pdu import PDU, EthernetFrame
-
-# Create a PDU
-pdu = PDU(type=0x01, duration=500, payload=b'\x01\x02\x03\x04')
-encoded_pdu = pdu.encode()
-decoded_pdu = PDU.decode(encoded_pdu)
-
-# Create an Ethernet Frame
-frame = EthernetFrame(
-    dstaddr=(0x01, 0x02, 0x03, 0x04, 0x05, 0x06),
-    srcaddr=(0x11, 0x12, 0x13, 0x14, 0x15, 0x16),
-    ethtype=0x0800,
-    payload=bytes([0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08])
+from ProtocolDataUnits.pdu import create_pdu_format, PDU
+
+# Create a simple PDU format
+my_pdu_format = create_pdu_format(
+    24, 'big',
+    ('uint8', 'type'),
+    ('float', 'value1'),
+    ('double', 'value2')
+)
```

Comment:
- Would it make sense to have constants defined for allowed types and endianness?
- `bigendian` might be more descriptive than just `big`
- Can't you infer size from the rest of the fields, if not specified? What is the expected behavior if the size does not match the sum of lengths of all the fields?

Reply:
Agreed, What do you think about these below changes @mchitre ?

### 1. **Constants for Allowed Types and Endianness**:
   - Define constants for data types and endianness. This makes the code more readable and reduces the chance of errors when using the types or endianness in different places.

   Example:
   ```python
   # Constants for byte order
   BIG_ENDIAN = '>'
   LITTLE_ENDIAN = '<'
   
   # Constants for field types
   UINT8 = 'uint8'
   INT8 = 'int8'
   UINT16 = 'uint16'
   INT16 = 'int16'
   UINT32 = 'uint32'
   INT32 = 'int32'
   UINT64 = 'uint64'
   INT64 = 'int64'
   FLOAT = 'float'
   DOUBLE = 'double'
   ```

And Update the existing code to use these constants. For example, instead of `'big'`, use `BIG_ENDIAN`, and for data types, use `UINT8`, `FLOAT`, etc.

   ```python
   def order(self, byte_order):
       self.byte_order = BIG_ENDIAN if byte_order == 'bigendian' else LITTLE_ENDIAN
       return self
   ```

### 2. **More Descriptive Endianness**:
   Change the description of endianness from `big` to `bigendian` and `little` to `littleendian` in both the code and documentation.

   Update the byte order checks:
   ```python
   def order(self, byte_order):
       self.byte_order = BIG_ENDIAN if byte_order == 'bigendian' else LITTLE_ENDIAN
       return self
   ```

### 3. **Inferring Size from Fields**:
   If the PDU size is not provided, infer the size from the sum of field lengths.

   - First, create a mapping of field types to their respective sizes (using `struct.calcsize`).

     ```python
     FIELD_SIZES = {
         UINT8: struct.calcsize('B'),
         INT8: struct.calcsize('b'),
         UINT16: struct.calcsize('H'),
         INT16: struct.calcsize('h'),
         UINT32: struct.calcsize('I'),
         INT32: struct.calcsize('i'),
         UINT64: struct.calcsize('Q'),
         INT64: struct.calcsize('q'),
         FLOAT: struct.calcsize('f'),
         DOUBLE: struct.calcsize('d'),
     }
     ```

   - Implement logic in the `encode` method to sum the sizes of all fields if the length is not specified:

     ```python
     def length(self, length=None):
         if length is None:
             self.pdu_length = sum(FIELD_SIZES[field_type] for field_type, name, value in self.fields)
         else:
             self.pdu_length = length
         return self
     ```

### 4. **Size Mismatch Behavior**:
   If the user specifies a size that does not match the sum of the field sizes, raise an error to notify them.

   Example:
   ```python
   def encode(self, data, compress=False):
       # Infer the PDU length if not set
       if self.pdu_length is None:
           self.length()

       total_field_size = sum(FIELD_SIZES[field_type] for field_type, name, value in self.fields)
       if total_field_size != self.pdu_length:
           raise ValueError(f"Specified PDU length {self.pdu_length} does not match total field length {total_field_size}")
       # Continue encoding logic...
   ```

Reply:
Looks great!